output/cloud/expv2/flush.go

```diff
@@ -43,60 +46,110 @@ func (f *metricsFlusher) flush() error {
 	// the metricSetBuilder is used for doing it during the traverse of the buckets.
 
 	var (
-		seriesCount  int
-		batchesCount int
-		start        = time.Now()
+		start       = time.Now()
+		batches     []*pbcloud.MetricSet
+		seriesCount int
 	)
 
 	defer func() {
 		f.logger.
 			WithField("t", time.Since(start)).
 			WithField("series", seriesCount).
 			WithField("buckets", len(buckets)).
-			WithField("batches", batchesCount).Debug("Flush the queued buckets")
+			WithField("batches", len(batches)).Debug("Flush the queued buckets")
 	}()
 
 	msb := newMetricSetBuilder(f.testRunID, f.aggregationPeriodInSeconds)
 	for i := 0; i < len(buckets); i++ {
 		for timeSeries, sink := range buckets[i].Sinks {
 			msb.addTimeSeries(buckets[i].Time, timeSeries, sink)
-			if len(msb.seriesIndex) < f.maxSeriesInBatch {
+			if len(msb.seriesIndex) <= f.maxSeriesInBatch {
 				continue
 			}
 
-			// we hit the batch size, let's flush
-			batchesCount++
+			// We hit the batch size, let's flush
 			seriesCount += len(msb.seriesIndex)
-			if err := f.push(msb); err != nil {
-				return err
-			}
+			batches = append(batches, msb.MetricSet)
+			f.reportDiscardedLabels(msb.discardedLabels)
+
+			// Reset the builder
 			msb = newMetricSetBuilder(f.testRunID, f.aggregationPeriodInSeconds)
 		}
 	}
 
-	if len(msb.seriesIndex) < 1 {
-		return nil
-	}
-
 	// send the last (or the unique) MetricSet chunk to the remote service
-	batchesCount++
 	seriesCount += len(msb.seriesIndex)
-	return f.push(msb)
+	batches = append(batches, msb.MetricSet)
+	f.reportDiscardedLabels(msb.discardedLabels)
+
+	return f.flushBatches(batches)
+}
+
+func (f *metricsFlusher) flushBatches(batches []*pbcloud.MetricSet) error {
+	var (
+		wg   = sync.WaitGroup{}
+		errs = make(chan error)
+		done = make(chan struct{})
+		stop = make(chan struct{})
+
+		workers   = int(math.Min(float64(len(batches)), float64(f.batchPushConcurrency)))
+		chunkSize = len(batches) / workers
+	)
+
+	wg.Add(workers)
+	for workersIndex := 0; workersIndex < workers; workersIndex++ {
+		offset := (workersIndex * chunkSize)
+		end := offset + chunkSize
+		if workersIndex == workers-1 {
+			end = len(batches)
```

Comment:
I would not do it like that ... and I m not certain this math works out perfectly :grimacing: 

I would've just given each worker a channel and written to it from the main goroutine adding all the pieces and let the workers get them on at a time.

This way they will balance out if one piece/worker takes longer and not hard math needed.

Reply:
Also at least in theory this can be done while the pieces are being broken down.